File: next/src/utils/MazeGenerator.ts

export enum Direction {
  North = 1,
  East = 2,
  South = 4,
  West = 8,
}

export type Cell = {
  walls: Direction;
  x: number;
  y: number;
  visited: boolean;
  //todo warm lights + trapdoors
}

export type Maze = {
  cells: Cell[][];
  position: { x: number, y: number, direction: Direction }
}

export function generateMaze(size: number): Maze {
  const cells: Cell[][] = [];
  for (var x = 0; x < size; x++) {
    cells.push([...Array(size)].map((_, y) => ({ walls: Direction.North | Direction.East | Direction.South | Direction.West, visited: false, x: y, y: x } as Cell)));
  }

  const stack: Cell[] = [];
  var x = rnd(size);
  var y = rnd(size);
  let first = cells[y][x];
  first.visited = true;
  stack.push(first);

  while (stack.length > 0) {
    let current = stack.pop();
    if (current == undefined) break;

    let neighbors = getNeighbors(current, cells);
    if (neighbors.length > 0) {
      var neighbor = neighbors[Math.floor(Math.random() * neighbors.length)];
      stack.push(current);
      if (neighbor.x > current.x) {
        neighbor.walls &= ~Direction.West;
        current.walls &= ~Direction.East;
      } else if (neighbor.x < current.x) {
        neighbor.walls &= ~Direction.East;
        current.walls &= ~Direction.West;
      } else if (neighbor.y > current.y) {
        neighbor.walls &= ~Direction.North;
        current.walls &= ~Direction.South;
      } else if (neighbor.y < current.y) {
        neighbor.walls &= ~Direction.South;
        current.walls &= ~Direction.North;
      }
      neighbor.visited = true;
      stack.push(neighbor);
    }
  }

  for (var x = 0; x < cells.length; x++) {
    for (var y = 0; y < cells[x].length; y++) {
      cells[y][x].visited = false;
    }
  }

  let position = { x: rnd(size), y: rnd(size), direction: Direction.North };
  cells[position.y][position.x].visited = true;
  return { cells, position } as Maze;
}

function getNeighbors(cell: Cell, maze: Cell[][]) {
  var x = cell.x;
  var y = cell.y;
  var neighbors = [];
  if (x > 0) neighbors.push(maze[y][x - 1]);
  if (x < maze[0].length - 1) neighbors.push(maze[y][x + 1]);
  if (y > 0) neighbors.push(maze[y - 1][x]);
  if (y < maze.length - 1) neighbors.push(maze[y + 1][x]);
  return neighbors.filter(c => !c.visited);
}

function rnd(max: number) {
  return Math.floor(Math.random() * max);
}

File: NetCoreGameServer/NetCoreGameServer/Service/MazeGenerator.cs
using System.Drawing;
using NetCoreGameServer.Data.Model;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace NetCoreGameServer.Service;

public class MazeGenerator
{
    private static readonly Random r = new();

    public static Maze Generate(int size)
    {
        var cells = new Cell[size][];
        for (var X = 0; X < size; X++)
        {
            cells[X] = Enumerable.Range(0, size).Select((_, y) => new Cell
            {
                Walls = Direction.North | Direction.East | Direction.South | Direction.West,
                Visited = false,
                X = y,
                Y = X
            }).ToArray();
        }

        Stack<Cell> stack = new Stack<Cell>();
        var x = r.Next(size);
        var y = r.Next(size);
        var first = cells[y][x];
        first.Visited = true;
        stack.Push(first);

        while (stack.Count > 0)
        {
            var current = stack.Pop();
            if (current == null) break;

            var neighbors = GetNeighbors(current, cells);
            if (neighbors.Length > 0)
            {
                var neighbor = neighbors[r.Next(neighbors.Length-1)];
                stack.Push(current);
                if (neighbor.X > current.X)
                {
                    neighbor.Walls &= ~Direction.West;
                    current.Walls &= ~Direction.East;
                }
                else if (neighbor.X < current.X)
                {
                    neighbor.Walls &= ~Direction.East;
                    current.Walls &= ~Direction.West;
                }
                else if (neighbor.Y > current.Y)
                {
                    neighbor.Walls &= ~Direction.North;
                    current.Walls &= ~Direction.South;
                }
                else if (neighbor.Y < current.Y)
                {
                    neighbor.Walls &= ~Direction.South;
                    current.Walls &= ~Direction.North;
                }

                neighbor.Visited = true;
                stack.Push(neighbor);
            }
        }

        for (var x2 = 0; x2 < cells.Length; x2++)
        {
            for (var y2 = 0; y2 < cells[x2].Length; y2++)
            {
                cells[y2][x2].Visited = false;
            }
        }
        
        var position = new Position
        {
            X = r.Next(size - 1),
            Y = r.Next(size - 1),
            Direction = Direction.North
        };
        cells[position.Y][position.X].Visited = true;

        return new Maze
        {
            Cells = cells,
            Position = position
        };
    }

    private static Cell[] GetNeighbors(Cell cell, Cell[][] maze)
    {
        var x = cell.X;
        var y = cell.Y;
        var neighbors = new List<Cell>();
        if (x > 0) neighbors.Add(maze[y][x - 1]);
        if (x < maze[0].Length - 1) neighbors.Add(maze[y][x + 1]);
        if (y > 0) neighbors.Add(maze[y - 1][x]);
        if (y < maze.Length - 1) neighbors.Add(maze[y + 1][x]);
        return neighbors.Where(c => !c.Visited).ToArray();
    }
}